Handle SQLRecoverableException. Expose remove abandoned properties

Lorren opened the pull request:
Minor code changes to handle the sqlrecoverableexception in databaseutils, and expose properties on the realm configuration to periodically test and recover abandoned connections in the pool.

Lorren committed:
Handle SQLRecoverableException. Expose remove abandoned properties

imesh commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -598,13 +632,42 @@ private static void closeStatement(PreparedStatement preparedStatement) {
+        		log.warn("Couldn't recover from SQLRecoverableException - cause: " + reason);
+        	}
+        } catch (SQLException e) {
+            log.error("Error occurred during SQLRecoverableException handling." + e.getMessage());
```
It would be better to log the exception object without concatenating the message to the error log, otherwise the stack-trace would not be visible:
log.error("Error occurred during SQLRecoverableException handling.", e);

Lorren replied:
Sorry, I was copying from the class pattern and missed passing the exception to the logger.  Thanks for catching this.

imesh commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -598,13 +632,42 @@ private static void closeStatement(PreparedStatement preparedStatement) {
+    	}
+    }
+    
+    	private static void handleSQLRecoverableException(Connection dbConnection, SQLRecoverableException ex) {
```
It seems like handleSQLRecoverableException() method is trying to handle a recoverable situation of a close() method invocation. IMO if the connection is already closed we might not need to handle it again. Can you please explain your intension here?

Lorren replied:
The connection is closing unexpectedly, without a clear indication as to why.  
When this event occurs the connection is left in an unknown state -- possibly abandoned.  
Rather than swallowing the exception, this attempts to close it cleanly.  
If left in the unknown state the connection may be unusable but still occupy a slot in the pool.  This connection may be served up to another request for resources.  If it isn't usable, it will again be returned to the pool and another request will be issued for a resource, This incurs another round trip which may, again get an unusable connection.  With the addition of each unusable connection in the pool, the performance will degrade.

Another change in this class exposes the logAbandoned property.  Implementing and enabling this property will indicate whether or not the connections are, in fact, unusable and the root cause will be indicated in the stack trace. 

Note: "Logging of abandoned Statements and Connections adds overhead for every Connection open or new Statement because a stack trace has to be generated."

Therefore, after addressing the root cause and eliminating the issue, disable this property.

imesh commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -574,7 +602,11 @@ public static void closeConnection(Connection dbConnection) {
 
         if (dbConnection != null) {
             try {
+				if (dbConnection != null && !dbConnection.isClosed()) {
```
It would be better to fix the indentation issues, can see in couple of places.

Lorren replied:
Fixed. The compare tool didn't copy over the whitespace nicely.

Lorren committed:
fix spaces and log messages

imesh commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -598,13 +636,42 @@ private static void closeStatement(PreparedStatement preparedStatement) {
+        log.info("Attempting recovery from thrown SQLRecoverableException ..." + ex.getLocalizedMessage(), ex);
+        try {
+        	if(dbConnection != null){
+        		dbConnection.close();
```
Indentation seems to be not correct.

imesh commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -598,13 +636,42 @@ private static void closeStatement(PreparedStatement preparedStatement) {
+    	}
+    }
+    
+    private static void handleSQLRecoverableException(Connection dbConnection, SQLRecoverableException ex) {
```
Method name handleSQLRecoverableException() does not imply that it is trying to recover a close() operation. On the other hand, would it require to retry a database connection close operation? Would you mind explaining your intention here?

imesh commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -165,6 +166,22 @@ public static DataSource createUserStoreDataSource(RealmConfiguration realmConfi
             poolProperties.setValidationInterval(DEFAULT_VALIDATION_INTERVAL);
         }
+		
+        if (realmConfig.getUserStoreProperty(JDBCRealmConstants.REMOVE_ABANDONED) != null && 
```
It might be better to use StringUtils.isNotEmpty() here.

imesh commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -574,7 +606,11 @@ public static void closeConnection(Connection dbConnection) {
         if (dbConnection != null) {
             try {
-                dbConnection.close();
+				if (dbConnection != null && !dbConnection.isClosed()) {
```
Indentation seems to be not correct.

blue666man replied:
On line 609, "dbConnection != null" is redundant from line 608 and should be removed.

We should actually do 2 checks here.  Refer to [DBCP 1.4 API docs](http://commons.apache.org/proper/commons-dbcp/api-1.4/org/apache/commons/dbcp/BasicDataSource.html): 
1. if (datasource.isClosed()) // If true, this data source is closed and no more connections can be retrieved from this datasource.  We need to check this to see if its even possible to gain any more connections from the JDBC DataSource object.  So, call createRealmDataSource() above again with "datasource = createRealmDataSource(realmConfig);" and retry this whole method again (refactor to an overloaded exception handler method)
2. "else if (!dbConnection.isClosed())" // as is in the PR call dbConnection.close()"

imesh commented:
If we were to handle SQLRecoverableExceptions, we may need to handle both connection.close() and prepareStmt.executeQuery() operations. Otherwise this solution would not work in all scenarios.

blue666man commented:
Some overall comments and points here, mostly pertaining to Java 7 try-with-resources as well as connection pooling with DBCP:
1. Overall goal of this PR should not only be handling SQLRecoverableExceptions, but to fully and automatically handle all closeable objects:  DataSource, Connection, PreparedStatement / Statement, and ResultSet.
2. All methods should be changed to the Java 7 try-with-resources pattern.  Basically this is looks like
   `try (Connection conn = ...; Statement stmt = ...;) {
   try (ResultSet rs = ...;) {
     // Do things with rs
   } catch (SQLRecoverableException sqre) { // Issues with rs itself, not stmt or conn
     //call recoverSqlException(sqre, stmt) // This should recover the conn and stmt at the least and retry the execution of the statement
   } catch (SQLException sqex) { // Issues with rs itself, not stmt or conn
     // log this unrecoverable exception
     // set final String errorMessage to something meaningful.  Include the fact that this error is limited to executing the statement / creating a ResultSet only and that the connection and statement object themselves are ok
     throw new UserStoreException(errorMessage, sqex);
   }  // finally clause omitted because of try-with-resources
   } catch  (SQLRecoverableException sqre) { // Issues with stmt or conn
     //call recoverSqlException(sqre, stmt, conn) // This should recover the conn if possible, if not close it, and if it can't close it, null it and get a new connection from the datasource, and if getting a new connection fails, recreate the datasource from createUserStoreDataSource().  Small protected scope methods should be used here.
   } catch (SQLException sqex) { // Issues with stmt or conn
     // log this unrecoverable exception
     // set final String errorMessage to something meaningful.  Include the fact that this error not recoverable, and that both the DBCP configuration (Carbon DataSource) as well as the database's data model should be checked.
     throw new UserStoreException(errorMessage, sqex);
   }  // finally clause omitted because of try-with-resources`

Lorren committed:
additional sqlrecoverabelexception handlers, empty property checking

Lorren committed:
additional sqlrecoverabelexception handlers, empty property checking

Lorren committed:
Merge remote-tracking branch 'origin/sqlrecoverable-4.4.x' into sqlrecoverable-4.4.x

Conflicts:
	core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

Lorren committed:
Merge remote-tracking branch 'origin/sqlrecoverable-4.4.x' into sqlrecoverable-4.4.x

Lorren committed:
Merge remote-tracking branch 'origin/sqlrecoverable-4.4.x' into sqlrecoverable-4.4.x

Lorren committed:
Use try-with-resources

Lorren committed:
merge sqlrecoverable-4.4.x

Conflicts:
	core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

Lorren committed:
use interface in handleSQLRecoverable

Lorren committed:
use interface in handlesqlrecoverable method call.  add unit test to repository

Lorren committed:
merge sqlrecoverable-4.4.x --> 4.4.x

Lorren committed:
Java7 multi-catch syntax . use generic argument types for to use interfaces to remove code duplication.

Lorren committed:
merge sqlrecoverable-4.4.x --> 4.4.x

blue666man commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -262,35 +277,33 @@ private static DataSource createRealmDataSource(RealmConfiguration realmConfig)
             }
             throw new UserStoreException(errorMessage, e);
-        } finally {
-            DatabaseUtil.closeAllConnections(null, rs, prepStmt);
```
The try-with-resource pattern automatically handles the ResultSet and PreparedStatement objects, but the Connection, "dbConnection" in this scope, still needs to be closed with "DatabaseUtil.closeAllConnections(dbConnection);"

Lorren replied:
added finally {DatabaseUtil.close(dbconnection)}
rs, prepstmt are closed when they go out of scope.

blue666man commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -303,38 +316,35 @@ private static DataSource createRealmDataSource(RealmConfiguration realmConfig)
             }
             throw new UserStoreException(errorMessage, e);
-        } finally {
-            DatabaseUtil.closeAllConnections(null, rs, prepStmt);
```
The try-with-resource pattern automatically handles the ResultSet and PreparedStatement objects, but the Connection, "dbConnection" in this scope, still needs to be closed with "DatabaseUtil.closeAllConnections(dbConnection);"

Lorren replied:
done: DatabaseUtil.close(dbConnection)

blue666man commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -347,28 +357,25 @@ public static int getIntegerValueFromDatabase(Connection dbConnection, String sq
             }
             throw new UserStoreException(errorMessage, e);
-        } finally {
-            DatabaseUtil.closeAllConnections(null, rs, prepStmt);
```
The try-with-resource pattern automatically handles the ResultSet and PreparedStatement objects, but the Connection, "dbConnection" in this scope, still needs to be closed with "DatabaseUtil.closeAllConnections(dbConnection);"

Lorren replied:
done: DatabaseUtil.close(dbConnection)

blue666man commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -419,22 +426,19 @@ public static void udpateUserRoleMappingInBatchModeForInternalRoles(Connection d
-        } finally {
+        } finally {  // can remove since this is not needed with try-with-resources -
             if (localConnection) {
                 DatabaseUtil.closeAllConnections(dbConnection);
```
The try-with-resource pattern automatically handles the ResultSet and PreparedStatement objects, but the Connection, "dbConnection" in this scope, still needs to be closed with "DatabaseUtil.closeAllConnections(dbConnection);"

blue666man commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -556,8 +557,6 @@ public static void updateDatabase(Connection dbConnection, String sqlStmt, Objec
                 log.debug(errorMessage, e);
             }
             throw new UserStoreException(errorMessage, e);
-        } finally {
```
The try-with-resource pattern automatically handles the ResultSet and PreparedStatement objects, but the Connection, "dbConnection" in this scope, still needs to be closed with "DatabaseUtil.closeAllConnections(dbConnection);"

Lorren replied:
done: DatabaseUtil.close(dbConnection)

blue666man commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -262,35 +277,33 @@ private static DataSource createRealmDataSource(RealmConfiguration realmConfig)
+                    values = lst.toArray(new String[lst.size()]);
+                }
+                return values;
             }
```
Missing a catch statement for handling ResultSet issues; resultSet.next() can throw SQLRecoverableException as well as SQLException

Lorren replied:
resultset is not in scope outside the code block, so it isn't available to call or pass into a close statement.  Tried putting in a catch statement for the sqlrecoverable exception but that threw an exception on missing return values.

blue666man commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -262,35 +277,33 @@ private static DataSource createRealmDataSource(RealmConfiguration realmConfig)
+                return values;
             }
-            return values;
         } catch (SQLException e) {
```
How about multi-catch for SQLRecoverableExceptions?

blue666man commented on core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java:
```diff
@@ -571,71 +570,116 @@ public static Connection getDBConnection(DataSource dataSource) throws SQLExcept
-                preparedStatement.close();
-            } catch (SQLException e) {
-                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
+    private static void handleSQLRecoverableException(AutoCloseable dbObject, SQLRecoverableException ex, boolean retry){
```
Great stuff.  Maybe we can refactor line 72 in closeDatabasePoolConnection() to use this as well since DataSource in JDBC 7 is also an AutoCloseable object?

blue666man commented:
Overall observation; Apache DBCP, especially the old 1.x version used by WSO2, could be the source of the problems:

https://github.com/brettwooldridge/HikariCP/wiki/Bad-Behavior:-Handling-Database-Down

Lorren committed:
Java7 multi-catch syntax . use generic argument types for to use interfaces to remove code duplication.

Lorren committed:
Java7 multi-catch syntax . use generic argument types for to use interfaces to remove code duplication.

Lorren committed:
merge changes sqlrecoverable-4.4.x --> 4.4.x

Lorren commented:
moving to new pull request
